### ISSUE TITLE
Since 90ea539 cant properly close the popup terminal window.

### ISSUE THREAD

**Issue by m1t0z** [voldikss/vim-floaterm#246]
Hi!

First of all, let me thank you for such a wonderful plugin. 
After I've discovered it I'm constantly using it as a replacement of the built-in vim terminal in the separate split-buffer.

Unfortunately since 90ea539 commit (and at least till 3446cd0)  I'm not able to close the popup terminal window. All of the followings do not have the desired effect:
- `FloatermHide`
- `FloatermToggle`
- type `exit` inside the terminal and then Ctr-w, Ctr-q (however this approach sometimes works)

The last revision when everything works as expected is 5b0d6b5.

My env:
- vim version: VIM - Vi IMproved 8.2 (Included patches: 1-728)
- shell: fish

Regards,
Dmitry

**Comment by voldikss** [comment]
I've tried but can not reproduce it. Could you give regular steps for reproducing?

**Comment by voldikss** [comment]
Also a gif would be better

**Comment by m1t0z** [comment]
Hi @voldikss.

Here is a .gif which shows the difference between `90ea539` (nok) and `5b0d6b5` (ok) revisions on my machine:

![issue](https://user-images.githubusercontent.com/37900444/107185528-b20b8180-69f3-11eb-85c7-f8fb5575281e.gif)

**Comment by voldikss** [comment]
@m1t0z Sadly I still can not reproduce it. 90ea539 shouldn't affect the plugin behavior if using on vim8.

The plugin works as expected on my laptop. Maybe you should try with a minimal vimrc

![issue](https://user-images.githubusercontent.com/20282795/107189068-b2bc0c00-6a23-11eb-8b05-7de010e9ae3f.gif)

**Comment by voldikss** [comment]
Closing this issue. But still feel free to comment.

**Comment by m1t0z** [comment]
Hi @voldikss.

Unfortunately still have the same issue with the fresh vim-floaterm master (more precisely with a4c834bc).
Have unsuccessfully tried the following things:
- empty .vimrc config (only with vim-plug + vim-floaterm)
- use alternative terminal (my default is Gnome Terminal 3.36.2, have tried with Konsole)
- set `g:floaterm_shell` to `bash` (as my default shell is fish)
- set `shell` to `bash` (as my default shell is fish)

Nothing from those has helped me with my issue. Behavior is still the same - i'm not able to close the floaterm window.

Without deep understanding i've looked into "problem" 90ea539 revision and tried my luck by "kinda reverting" the change inside the `floaterm#window#hide` function.

Before:
```
function! floaterm#window#hide(bufnr) abort                                                                                                                                                                                                                               
  ...
  if has('nvim')
  ...
  else
    if s:winexists(winid)
      try 
        call popup_close(winid)
   ...
```

After:
```
function! floaterm#window#hide(bufnr) abort                                                                                                                                                                                                                               
  ...
  if has('nvim')
  ...
  else
    if s:winexists(winid) || 1     " <== IGNORING THE CHECK OF WINDOW EXISTENCE 
      try 
        call popup_close(winid)
   ...
```
And now it works for me :)

I didn't have time to look deeper to find the root-cause of the issue, but at least i'm happy to find the temporal workaround that works for me personally.

Do you have any ideas why `s:winexists(winid)` could result to false in my case?


Regards,
Dmitry

**Comment by voldikss** [comment]
Could I try logging in to your PC via something like teamviewer and try solving this issue?

**Comment by voldikss** [comment]
I feel it was caused by your vim version...

**Comment by m1t0z** [comment]
> I feel it was caused by your vim version...

Hmm... looks like it. 

I have zero experience in vim-plug-development but after the brief look into the code i've found something kinda-strange. 

In my env the built-in vim-function `getwininfo` when called without any parameters always return only 1-element-list with winid=1000. I suppose it is the main (or how its called) vim-window itself. 

What looks really strange that after creating a new popup window with another vim-builtin `popup_create` (in your code this is done inside `s:open_popup`) the `getwininfo` still returns only 1 element, thus not reflecting a newly created popup window.

Documentation of the `popup_create` says the following
```
 Returns a window-ID, which can be used with other popup functions. 
```

Here i'm not really sure how to interpret the last part of the sentence.
Does it means that the returned window-id can be used **only** with popup functions. And if so does `getwininfo` belongs to such popup functions. Could it be that by design window-id of popup-functions is not the same as window-id of other functions like `getwininfo`. Or maybe it is really a bug in my vim-version. Or some broken env setup at my site. Hm...don't know :)

Regards,
Dmitry.

**Comment by voldikss** [comment]
> In my env the built-in vim-function getwininfo when called without any parameters always return only 1-element-list with winid=1000. I suppose it is the main (or how its called) vim-window itself.

Yes.

> What looks really strange that after creating a new popup window with another vim-builtin popup_create (in your code this is done inside s:open_popup) the getwininfo still returns only 1 element, thus not reflecting a newly created popup window.

You should try `echom getwininfo(win_getid())`

> Does it means that the returned window-id can be used only with popup functions. 

No. It means the returned `winid` can be passed into other popup functions such as  `popup_hide()`, `popup_getpos()`, etc.

> And if so does getwininfo belongs to such popup functions. 

It's just a function. All functions are equal.

>  Could it be that by design window-id of popup-functions is not the same as window-id of other functions like getwininfo.

If we call `getwininfo()` without a specific winid parameter, it will return all windows in all tabpages but excluding popup windows. Hence getwininfo still returns only 1 element even if you have opened a popup window. You should pass it a window-ID of the popup window (which can be acquired through `win_getid()`) instead.

**Comment by m1t0z** [comment]
> If we call getwininfo() without a specific winid parameter, it will return all windows in all tabpages but excluding popup windows. Hence getwininfo still returns only 1 element even if you have opened a popup window. You should pass it a window-ID of the popup window (which can be acquired through win_getid()) instead.

Is it some vim implementation-internal knowledge? In the documentation of the `getwininfo` or `popup_create` i didn't found anything that will say that `getwininfo` when called without specific window-id will return information about all windows **except** popup-windows.

> You should try echom getwininfo(win_getid())

Tried this and several other things -> result is always the same. `getwininfo` when called with id of the popup_window always returns an empty dict (i.e. `[]`) :(

As experiment i've tried to call `popup_getoptions` with id of the popup_window and this successfully returns a full dict of popup-window options.

I have also briefly looked in the vim-internal implementation of the `f_getwininfo` ([link](https://github.com/vim/vim/blob/37487e16da7877129edee8d11b9b7f5c8df312c6/src/evalwindow.c#L497)). At least it looks like it should handle popup windows (when FEAT_PROP_POPUP is activated, which is true for my vim local vim installation).
However i didn't have enough time to dig in into that code. 

If you don't have any further ideas i will create an issue-question to vim github directly to ask if it is ok or not to receive an empty dictionary as result of calling `getwininfo` for popup_window id.

Regards,
Dmitry

**Comment by voldikss** [comment]
> except popup-windows.

Yes the documentation doesn't cover it. It's infered by me.

**Comment by voldikss** [comment]
> Tried this and several other things -> result is always the same. getwininfo when called with id of the popup_window always returns an empty dict (i.e. []) :(

Have you tried the latest vim version?

> I have also briefly looked in the vim-internal implementation of the f_getwininfo (link). At least it looks like it should handle popup windows (when FEAT_PROP_POPUP is activated, which is true for my vim local vim installation).
However i didn't have enough time to dig in into that code.

Yes, that's true.

> If you don't have any further ideas i will create an issue-question to vim github directly to ask if it is ok or not to receive an empty dictionary as result of calling getwininfo for popup_window id. 

I don't have the issue on my vim. Maybe you had better upgrade your vim to the latest version first.

**Comment by voldikss** [comment]
>  At least it looks like it should handle popup windows (when FEAT_PROP_POPUP is activated, which is true for my vim local vim installation).

Look at [line534](https://github.com/vim/vim/blob/37487e16da7877129edee8d11b9b7f5c8df312c6/src/evalwindow.c#L534). I am not sure but I guess that if we call `getwininfo()` without passing it a `winid` argument. the `wp`("window pointer" I guess ) would be `NULL`, which means popup windows info would be handled.

**Comment by m1t0z** [comment]
> I don't have the issue on my vim. Maybe you had better upgrade your vim to the latest version first.

And this was it :)

I've built&install the last available vim tag version (v8.2.2580) and the issue has gone :+1: 

Now it behaves exactly the way you were describing, i.e. call of the `getwininfo()` without any arguments returns the list of all windows except popups and call of the `getwininfo()` with id of the popup-window returns information about this popup window (and not an empty dict as it was before).

So i believe you could close this issue. Hopefully the vim version i've used before is not common among other users of your amazing plugin. 

Huge thanks for your support and best regards,
Dmitry

**Comment by voldikss** [comment]
Nice to hear that.
